core/src/main/scala/org/bykn/bosatsu/LetFree.scala

```diff
@@ -129,42 +180,69 @@ object LetFreeConversion {
   type LetFreePM = PackageMap.Typed[(Declaration, LetFreeExpressionTag)]
   type LetFreePac = Package.Typed[(Declaration, LetFreeExpressionTag)]
 
-  type PatternEnv = Map[Int, LetFreeExpression]
+  type PatternEnv[T] = Map[Int, T]
 
   sealed trait PatternMatch[+A]
   case class Matches[A](env: A) extends PatternMatch[A]
   case object NoMatch extends PatternMatch[Nothing]
   case object NotProvable extends PatternMatch[Nothing]
 
-  val noop: (LetFreeExpression, PatternEnv) => PatternMatch[PatternEnv] = { (_, env) => Matches(env) }
-  val neverMatch: (LetFreeExpression, PatternEnv) => PatternMatch[Nothing] = { (_, _) => NoMatch }
+  def noop[T]: (T, PatternEnv[T]) => PatternMatch[PatternEnv[T]] = { (_, env) => Matches(env) }
+  def neverMatch[T]: (T, PatternEnv[T]) => PatternMatch[Nothing] = { (_, _) => NoMatch }
 
-  def structListAsList(lfe: LetFreeExpression): PatternMatch[List[LetFreeExpression]] = {
+  def structListAsList(lfe: LetFreeExpression): Option[List[LetFreeExpression]] = {
     lfe match {
-      case LetFreeExpression.Struct(0, _) => Matches(Nil)
-      case LetFreeExpression.Struct(1, List(value, tail)) => structListAsList(tail) match {
-        case Matches(lst) => Matches(value :: lst)
-        case notMatch => notMatch
-      }
-      case _ => NotProvable
+      case LetFreeExpression.Struct(0, _, _) => Some(Nil)
+      case LetFreeExpression.Struct(1, List(value, tail), _) => structListAsList(tail).map(value :: _)
+      case _ => None
     }
   }
 
   def listAsStructList(lst: List[LetFreeExpression]): LetFreeExpression =
-    lst.foldRight(LetFreeExpression.Struct(0, Nil)) { case (lfe, acc) => LetFreeExpression.Struct(1, List(lfe, acc)) }
+    lst.foldRight(LetFreeExpression.Struct(0, Nil, DataFamily.Enum)) { case (lfe, acc) => LetFreeExpression.Struct(1, List(lfe, acc), DataFamily.Enum) }
+
+  case class LitValue(toAny: Any) {
+    def equivToLit(lit: Lit) = lit match {
+      case Lit.Integer(i) => i == toAny
+      case Lit.Str(s) => s == toAny
+    }
+  }
+
+  object LitValue {
+    def fromLit(lit: Lit) = lit match {
+      case Lit.Integer(i) => LitValue(i)
+      case Lit.Str(s) => LitValue(s)
+    }
+  }
 
-  private def maybeBind(pat: LetFreePattern): (LetFreeExpression, PatternEnv) => PatternMatch[PatternEnv] =
+  implicit val neToLitValue: LetFreeExpression => Option[LitValue] = {
+    case LetFreeExpression.Literal(lit) => Some(LitValue.fromLit(lit))
+    case _ => None
+  }
+  implicit val neToStruct: (LetFreeExpression, DataFamily) => Option[(Int, List[LetFreeExpression])] = {
+    case (LetFreeExpression.Struct(enum, args, _), df) => Some((enum, args))
+    case _ => None
+  }
+  implicit val neToList: LetFreeExpression => Option[List[LetFreeExpression]] = structListAsList(_)
+  implicit val neFromList: List[LetFreeExpression] => LetFreeExpression = listAsStructList(_)
+
+  def maybeBind[T](pat: LetFreePattern)(implicit
```

Comment:
I would at least consider what I call the "module pattern" here. Where you would do this:
```scala
abstract class MaybeBind[A](pat: LetFreePattern) {
  def toLitValue(t: A): Option[LitValue]
  def toStruct(t: A, df: DataFamily): Opiton[(Int, List[A])]
  ...

  // concrete final (and some private) methods below.
}
```